### PR TITLE
[rosetta] Add error message to rosetta fee estimation

### PR DESCRIPTION
**Pull request by gregnazario**
### Description
Had to refactor this because it was unwieldy to understand just to add one error case

Basically now, always simulate the transaction, and if the max gas provided by the input isn't enough, then error out.  Once we have better error messages in simulation, this weird hack won't be needed.

### Test Plan
CICD, I'll do some manual testing

- - -
This change is [Reviewable](https://reviewable.io/reviews/aptos-labs/aptos-core/4678)

### CODE REVIEW

File: crates/aptos-rosetta/src/construction.rs [hunk 1]
```diff
@@ -313,101 +288,160 @@ async fn construction_metadata(
...
+    Ok(internal_operation)
+}
+
+async fn simulate_transaction(
+    rest_client: &aptos_rest_client::Client,
+    chain_id: ChainId,
+    options: &MetadataOptions,
+    internal_operation: &InternalOperation,
+    sequence_number: u64,
+) -> ApiResult<(Amount, u64, u64)> {
```
Comment (banool) [inline crates/aptos-rosetta/src/construction.rs:300]: Nit: Return a struct or some newtypes, returning a bunch of the same type is a little spooky.

File: crates/aptos-rosetta/src/construction.rs [hunk 2]
```diff
@@ -313,101 +288,160 @@ async fn construction_metadata(
...
 
-            // If there's a gas unit price we're using it, max gas doesn't matter the API will overwrite it
-            if let Some(gas_unit_price) = request.options.gas_price_per_unit {
-                transaction_factory = transaction_factory.with_gas_unit_price(gas_unit_price.0)
-            }
+    // Sign the transaction with a dummy signature of all zeros as required by the API
+    let signed_transaction = SignedTransaction::new(
+        unsigned_transaction,
+        public_key,
+        Ed25519Signature::try_from([0u8; 64].as_ref()).expect("Zero signature should always work"),
```
Comment (banool) [inline crates/aptos-rosetta/src/construction.rs:339]: Usually we have a method for this kind of thing, do we not have one here? Like `Ed25519Signature::zero`.

File: crates/aptos-rosetta/src/error.rs [hunk 3]
```diff
@@ -21,7 +21,7 @@ pub enum ApiError {
     InvalidTransferOperations(Option<&'static str>),
     InvalidSignatureType,
     InvalidMaxGasFees,
-    MaxGasFeeTooLow,
+    MaxGasFeeTooLow(Option<String>),
```
Comment (banool) [inline crates/aptos-rosetta/src/error.rs:24]: I wonder why we include the string here if we don't use it below?

**Review by banool — APPROVED** [review]

File: crates/aptos-rosetta/src/error.rs (hunk 3 above)
Comment (gregnazario) [reply crates/aptos-rosetta/src/error.rs:24]: This is really annoying, but you can't have the "details" in the message by default when listing out the messages, it's a spec thing

**Comment by bowenyang007** [comment]
Did you do the manual testing?

**Review by bowenyang007 — APPROVED** [review]

**Pull request merged**